### PR TITLE
Edit: Allow editing the inline diff

### PR DESCRIPTION
**Pull request by umpox**
## Description

closes https://github.com/sourcegraph/cody/issues/4671

https://github.com/sourcegraph/cody/assets/9516420/7d468713-c928-461b-b0f0-192a91560d44

Problem:
- We auto-accept the task whenever there is _any_ change within the `selectionRange`.
   - This was primarily useful for tracking if the user immediately does "Undo" on a change, so we update the UI and clear out the diff.
- With the new inline diff view, doing this means that it's difficult for users' to modify Cody's response. As any change they make immediately removes all deleted lines and hide decorations.

Solution:
- Allow making edits within the selection range. We can ensure decorations do not expand so it's still clear exactly what edits Cody made.
- Track specifically for the edit being restored back to its original state. This will work to ensure we still capture when a user does an "Undo"

## Test plan

1. Make Edits (Generate, Edit, Test, Fix, Document)
2. Make changes to the diff
3. Check everything looks ok
4. Try to undo _some_ of the changes you made
5. Check the entire edit is not removed, as we're only undoing additional changes
6. Undo all the changes made after the edit
7. Check the entire edit is removed (e.g. codelens, injected whitespace, etc)

### CODE REVIEW

**Review by abeatrix — COMMENTED** [review]
I asked the LLM to update the function name but the whole function was shown as removed:

image


And when I tried to undo it, it broke my code 😓 
image

Looks like it might be Gemini issues? Should we disable it for Edit?

https://github.com/sourcegraph/cody/assets/68532117/6b59c248-e900-4fec-bdee-1a36afeaf855

**Comment by abeatrix** [comment]
unrelated to this PR but we should also update the ghost text not to show up in diff lines? 

image

**Review by abeatrix — APPROVED** [review]
Run into some issues when using the Gemini model, which seems to be unrelated to the changes in this PR. I was able to edit the green diff as described in the PR 👍

File: vscode/src/non-stop/FixupController.ts [hunk 1]
```diff
@@ -207,6 +207,56 @@ export class FixupController
         }
     }
 
+    /**
+     * Given a task, tracks upcoming changes in the associated document.
+     * If the change restores an applied task to its original state, we discard the task
+     * meaning any associated UI and behaviour is updated.
+     */
+    public registerDiscardOnRestoreListener(task: FixupTask): void {
```
Comment (abeatrix) [inline vscode/src/non-stop/FixupController.ts:215]: changelog?

**Comment by umpox** [comment]
Thanks @abeatrix I'll look into Gemini, maybe we should disable that for edit

**Pull request merged**